File: amino/runtime/matcher.py
"""Pattern matching and result handling."""

import dataclasses
from typing import Any


@dataclasses.dataclass
class MatchResult:
    id: Any
    matched: list[str] = dataclasses.field(default_factory=list)
    excluded: list[str] = dataclasses.field(default_factory=list)
    score: float | None = None
    warnings: list[str] = dataclasses.field(default_factory=list)


class Matcher:
    def __init__(self, config: dict[str, Any] | None = None):
        cfg = config or {}
        self._mode = cfg.get("mode", "all")
        self._key = cfg.get("key")
        self._order = cfg.get("order", "asc")
        self._aggregate = cfg.get("aggregate", "sum")
        self._threshold = cfg.get("threshold")

    def process(
        self,
        decision_id: Any,
        rule_results: list[tuple[Any, Any]],
        metadata: dict[Any, dict],
        warnings: list[str],
    ) -> MatchResult:
        if self._mode == "all":
            matched = [rid for rid, val in rule_results if val]
            return MatchResult(id=decision_id, matched=matched, warnings=list(warnings))

        if self._mode == "first":
            matched = [rid for rid, val in rule_results if val]
            if not matched:
                return MatchResult(id=decision_id, matched=[], warnings=list(warnings))
            if self._key:
                matched = sorted(
                    matched,
                    key=lambda rid: metadata.get(rid, {}).get(self._key, float("inf")),
                    reverse=(self._order == "desc"),
                )
            return MatchResult(id=decision_id, matched=[matched[0]], warnings=list(warnings))

        if self._mode == "inverse":
            excluded = [rid for rid, val in rule_results if not val]
            return MatchResult(id=decision_id, excluded=excluded, warnings=list(warnings))

        if self._mode == "score":
            total = 0.0
            for _rid, val in rule_results:
                if isinstance(val, bool):
                    total += 1.0 if val else 0.0
                elif isinstance(val, (int, float)):
                    total += float(val)
            result = MatchResult(id=decision_id, score=total, warnings=list(warnings))
            if self._threshold is not None and total >= self._threshold:
                result.matched = [rid for rid, val in rule_results if val]
            return result

        raise ValueError(f"Unknown match mode: {self._mode!r}")

File: go/matcher.go
package amino

import "sort"

// MatchConfig selects how rule results are aggregated into a MatchResult.
type MatchConfig struct {
	Mode      string   // "all" (default), "first", "inverse", "score"
	Key       string   // first: the Rule.Meta key to order by
	Order     string   // first: "asc" (default) or "desc"
	Aggregate string   // score: "sum" (the only aggregate)
	Threshold *float64 // score: when set and met, Matched lists the truthy rules
}

// MatchResult is the outcome of evaluating a rule set against one record.
type MatchResult struct {
	ID       any      // the record's "id" value, if any
	Matched  []string // rule ids that matched (all, first, score with threshold)
	Excluded []string // rule ids that did not match (inverse)
	Score    *float64 // the aggregate (score)
	Warnings []string // validation warnings in loose mode
}

type ruleResult struct {
	id    string
	value any
}

func (c *CompiledRules) match(id any, results []ruleResult, warnings []string) (MatchResult, error) {
	out := MatchResult{ID: id, Matched: []string{}, Excluded: []string{}, Warnings: warnings}
	if out.Warnings == nil {
		out.Warnings = []string{}
	}
	cfg := c.cfg
	mode := cfg.Mode
	if mode == "" {
		mode = "all"
	}
	switch mode {
	case "all":
		for _, r := range results {
			if truthy(r.value) {
				out.Matched = append(out.Matched, r.id)
			}
		}
	case "first":
		var matched []string
		for _, r := range results {
			if truthy(r.value) {
				matched = append(matched, r.id)
			}
		}
		if len(matched) == 0 {
			return out, nil
		}
		if cfg.Key != "" {
			desc := cfg.Order == "desc"
			keyOf := func(id string) (float64, bool) {
				meta := c.meta[id]
				if meta == nil {
					return 0, false
				}
				return toFloat(meta[cfg.Key])
			}
			sort.SliceStable(matched, func(i, j int) bool {
				a, aok := keyOf(matched[i])
				b, bok := keyOf(matched[j])
				if !aok {
					return false
				}
				if !bok {
					return true
				}
				if desc {
					return a > b
				}
				return a < b
			})
		}
		out.Matched = []string{matched[0]}
	case "inverse":
		for _, r := range results {
			if !truthy(r.value) {
				out.Excluded = append(out.Excluded, r.id)
			}
		}
	case "score":
		total := 0.0
		for _, r := range results {
			switch v := r.value.(type) {
			case bool:
				if v {
					total++
				}
			default:
				if f, ok := toFloat(v); ok {
					total += f
				}
			}
		}
		out.Score = &total
		if cfg.Threshold != nil && total >= *cfg.Threshold {
			for _, r := range results {
				if truthy(r.value) {
					out.Matched = append(out.Matched, r.id)
				}
			}
		}
	default:
		return out, newError(CodeConfig, "unknown match mode %q", mode)
	}
	return out, nil
}
